src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -350,6 +327,93 @@ UpdateRelationsToLocalShardTables(Node *node, List *relationShardList)
 }
 
 
+/*
+ * ReplaceRelationConstraintByShardConstraint replaces given OnConflictExpr's
+ * constraint id with constraint id of the corresponding shard.
+ */
+static bool
+ReplaceRelationConstraintByShardConstraint(List *relationShardList,
+										   OnConflictExpr *onConflict)
+{
+	Oid constraintId = onConflict->constraint;
+
+	if (!OidIsValid(constraintId))
+	{
+		return false;
+	}
+
+	Oid constraintRelationId = InvalidOid;
+
+	HeapTuple heapTuple = SearchSysCache1(CONSTROID, ObjectIdGetDatum(constraintId));
+	if (HeapTupleIsValid(heapTuple))
+	{
+		Form_pg_constraint contup = (Form_pg_constraint) GETSTRUCT(heapTuple);
+
+		constraintRelationId = contup->conrelid;
+		ReleaseSysCache(heapTuple);
+	}
+
+	/*
+	 * We can return here without calling the walker function, since we know there
+	 * will be no possible tables or constraints after this point, by the syntax.
+	 */
+	if (!OidIsValid(constraintRelationId))
+	{
+		ereport(ERROR, (errmsg("Invalid relation id (%u) for constraint: %s",
```

Comment:
a minor one:

I prefer to use `errcode(ERRCODE_INTERNAL_ERROR)` for such internal bug catches.
It it makes sense to you,  you need to do will be that:
```c
    ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
                    errmsg("Invalid relation id (%u) for constraint: %s",             
                            constraintRelationId,
                            get_constraint_name(constraintId))));
```

Reply:
I think for errors the default is already ERRCODE_INTERNAL_ERROR.

```
All else is optional.  errcode() defaults to
 * ERRCODE_INTERNAL_ERROR if elevel is ERROR or more, ERRCODE_WARNING
 * if elevel is WARNING, or ERRCODE_SUCCESSFUL_COMPLETION if elevel is
 * NOTICE or below.
```

Though it might be good to have that explicitly there :) But these are more ILLEGAL_STATE rather than internal errors. I mean they are internal errors but internal error is a general thing here. @onurctirtir do you know any error code that is like ILLEGAL_STATE? I checked but couldn't find anything :) So maybe we can stick to ERRCODE_INTERNAL_ERROR

Reply:
>I think for errors the default is already ERRCODE_INTERNAL_ERROR.

Hmm, I didn't know that, thanks :)
Then my comment can be ignored 

>do you know any error code that is like ILLEGAL_STATE?

Looked into that now but I couldn't find too